Add Browse() api to list items by folder

tamalsaha opened the pull request:
delimiter returns results in a directory-like fashion.
Results will contain only items whose names, aside from the
prefix, do not contain delimiter. Items whose names,
aside from the prefix, contain delimiter will have their name,
truncated after the delimiter, returned in prefixes.
Duplicate prefixes are omitted. Optional.

ref: https://godoc.org/cloud.google.com/go/storage#Query

This will allow us to build a file browser using Stow.

TamalSaha commented:
@matryer , I can't see which test is failing.

TamalSaha commented:
Actually this is not ready. To properly support delimiters, the concept of a "folder" is required.

matryer commented:
How come? Aren't strings and prefixes enough?

TamalSaha commented:
Hi,
Here is a code block that we wrote using Google Cloud SDK directly and looking to rewrite using Stow.
https://github.com/appscode/chartify/blob/upload-chart/pkg/repo/browse.go#L68

Essentially the issue is, when you call `Items()` on any prefix of a bucket, it may return a  new list of "prefixes" (direct subfolders) and "items" (direct files). So, currently it is not possible to represent these direct folder. We want to call Items() on these subjects again. How do you think it could be done?

TamalSaha commented:
@matryer, Also I wonder if this should be added as a separate `Browse()` api to avoid breaking any existing application.

matryer commented:
@TamalSaha can you spell out a real example using some made up paths:

E.g. 

```
/container/parent1/item1
/container/parent1/item2
/container/parent1/item3
/container/parent2/item4
/container/parent2/item5
/container/parent2/item6
```

sbward commented:
If I'm understanding this correctly, then the underlying issue is that it's impossible to distinguish file names from folders when the file name contains a delimiter.
If so, then maybe we should add an `IsDir()` method to the `Item` type.  @TamalSaha would such an addition resolve the issues you're having?

TamalSaha commented:
Here is an example. Say below is all the files in my bucket.
```yaml
/bucket/a/b/f1.txt
/bucket/a/b/f2.txt
/bucket/a/f3.txt
```

If I call `Items(prefix=/bucket/a, delimiter=/)`, Stow api should return:
```yaml
/bucket/a/b
/bucket/a/f3.txt
```

`/bucket/a/b` is not a "real" item/file. It is a prefix. Now, I can call `Items(prefix=/bucket/a/b, delimiter=/)`, to get
```yaml
/bucket/a/b/f1.txt
/bucket/a/b/f2.txt
```

Now, I can use this api to build a file browser. If `IsPrefix()` or `IsDir()` method is added to `Item`, it will solve the problem. I prefer the `IsPrefix()` since it does to introduce any new vocabulary in Stow.

sbward commented:
Ok, I think I understand now.  It seems that instead of showing all of the nested files under a prefix, you want to show only the files and prefixes which are immediately nested under a given prefix, so that you can provide a file browser experience similar to desktop file browsers directly from the Stow API.

I know for a fact that what @TamalSaha is suggesting is provided exactly as described by the Google Cloud Storage and Azure File Storage APIs.  It's probably a universal behavior at this point so we should check that and if so, add it to Stow.  The design that Google/Azure use is a combination of @TamalSaha's suggestion and my suggestion, that is, they have an *optional* delimiter argument which, if provided, replaces nested children with directory items.  I think we should probably do the same.

TamalSaha commented:
AWS api supports the same semantics; http://docs.aws.amazon.com/AmazonS3/latest/API/v2-RESTBucketGET.html

TamalSaha commented:
I am not well versed in OpenStack Swift api. But they seem to support a similar api:

https://developer.openstack.org/api-ref/object-storage/?expanded=show-container-details-and-list-objects-detail#show-container-details-and-list-objects

TamalSaha commented:
I have added a new Browse api:
`Browse(prefix, delimiter, cursor string, count int) (prefixes []string, items []Item, string, error)`

and updated `Items()` to implement in terms of `Browse()`. This means api stays backward compatible.

matryer commented on azure/container.go:
```diff
@@ -50,18 +50,25 @@ func (c *container) Item(id string) (stow.Item, error) {
 }
 
-func (c *container) Items(prefix, cursor string, count int) ([]stow.Item, string, error) {
+func (c *container) Browse(prefix, delimiter, cursor string, count int) ([]string, []stow.Item, string, error) {
```
I'm not keen on four return arguments, I wonder if we should consider wrapping it into a `Page` object or similar?

matryer replied:
Maybe we can abstract the concept of a page of items and reuse it for Items too?

TamalSaha replied:
"Maybe we can abstract the concept of a page of items and reuse it for Items too?" - Do you mind elaborating?

The idea of a page sounds reasonable.

matryer replied:
Something like this?

```
type ItemsPage struct {
  Prefix string
  Items []Item
  NextCursor string
  Len int
}
```

Would that give you everything you need?

TamalSaha replied:
I think so. But what does that Len mean?

tamalsaha commented:
@matryer , I have added `ItemPage` struct

tamalsaha commented:
It seems that tests are failing because of old Azure lib.
